src/crypto/CryptoNight_x86.h

```diff
@@ -412,46 +417,37 @@ inline void cryptonight_single_hash(const uint8_t *__restrict__ input, size_t si
     __m128i bx0 = _mm_set_epi64x(h0[3] ^ h0[7], h0[2] ^ h0[6]);
 
     uint64_t idx0 = h0[0] ^ h0[4];
-
-    for (size_t i = 0; i < ITERATIONS; i++) {
-        __m128i cx;
-
-        if (SOFT_AES) {
-            cx = soft_aesenc((uint32_t*)&l0[idx0 & MASK], _mm_set_epi64x(ah0, al0));
+    void* memoryPointer = ((uint8_t*) l0) + ((idx0) & MASK);
+
+    if(SOFT_AES && ALGO == xmrig::CRYPTONIGHT_HEAVY) {
+        for (size_t i = 0; i < ITERATIONS; i++) {
+            __m128i cx;
+            SINGLEHASH_LOOP_SOFTAES
+            SINGLEHASH_LOOP_COMMON
+            SINGLEHASH_LOOP_CNHEAVY
         }
-        else {  
-            cx = _mm_load_si128((__m128i *) &l0[idx0 & MASK]);
-            cx = _mm_aesenc_si128(cx, _mm_set_epi64x(ah0, al0));
+    } else if(!SOFT_AES && ALGO == xmrig::CRYPTONIGHT_HEAVY) {
+        for (size_t i = 0; i < ITERATIONS; i++) {
+            __m128i cx;
+            SINGLEHASH_LOOP_HARDAES
+            SINGLEHASH_LOOP_COMMON
+            SINGLEHASH_LOOP_CNHEAVY
         }
-        _mm_store_si128((__m128i *) &l0[idx0 & MASK], _mm_xor_si128(bx0, cx));
-        VARIANT1_1(&l0[idx0 & MASK]);
-        idx0 = EXTRACT64(cx);
-        bx0 = cx;
-
-        uint64_t hi, lo, cl, ch;
-        cl = ((uint64_t*) &l0[idx0 & MASK])[0];
-        ch = ((uint64_t*) &l0[idx0 & MASK])[1];
-        lo = __umul128(idx0, cl, &hi);
-
-        al0 += hi;
-        ah0 += lo;
+    } else {
+        for (size_t i = 0; i < ITERATIONS; i++) {
+            __m128i cx;
 
-        VARIANT1_2(ah0, 0);
-        ((uint64_t*)&l0[idx0 & MASK])[0] = al0;
-        ((uint64_t*)&l0[idx0 & MASK])[1] = ah0;
-        VARIANT1_2(ah0, 0);
-
-        ah0 ^= ch;
-        al0 ^= cl;
-        idx0 = al0;
+            if (SOFT_AES) {
+                SINGLEHASH_LOOP_SOFTAES
+            } else {  
+                SINGLEHASH_LOOP_HARDAES
+            }
 
-        if (ALGO == xmrig::CRYPTONIGHT_HEAVY) {
-            int64_t n  = ((int64_t*)&l0[idx0 & MASK])[0];
-            int32_t d  = ((int32_t*)&l0[idx0 & MASK])[2];
-            int64_t q = n / (d | 0x5);
+            SINGLEHASH_LOOP_COMMON
 
-            ((int64_t*)&l0[idx0 & MASK])[0] = n ^ q;
-            idx0 = d ^ q;
+            if (ALGO == xmrig::CRYPTONIGHT_HEAVY) {
```

Comment:
How is it suppose to happend here, since you're if & elseif took care of HEAVY already?
In a quatum SOFT_AES "may" state? (kidding)

Reply:
Ok you just changed it.

Reply:
I have to finish this patch, i'll commit last changes later.